File: app/src/main/java/com/ydoc/app/sync/LinkPreviewWorker.kt
package com.ydoc.app.sync

import android.content.Context
import androidx.work.BackoffPolicy
import androidx.work.Constraints
import androidx.work.CoroutineWorker
import androidx.work.ExistingWorkPolicy
import androidx.work.NetworkType
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.WorkManager
import androidx.work.WorkerParameters
import androidx.work.workDataOf
import com.ydoc.app.appContainer
import com.ydoc.app.logging.AppLogger
import com.ydoc.app.model.LinkPreview
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.flow.first

/**
 * 扫描 note 里的 URL，调 relay /api/links/preview 回填预览。
 * - 最多 5 条 URL，避免被单条笔记撑爆。
 * - 单条失败不影响其它；失败的 LinkPreview 也落库，UI 用 `error` 字段降级展示。
 * - 唯一 work 名 = "link_preview_<noteId>"，保证重复保存会 REPLACE 而不是堆队列。
 */
class LinkPreviewWorker(
    ctx: Context,
    params: WorkerParameters,
) : CoroutineWorker(ctx, params) {

    override suspend fun doWork(): Result {
        val noteId = inputData.getString(KEY_NOTE_ID) ?: return Result.failure()
        val container = applicationContext.appContainer
        val note = container.noteRepository.getNote(noteId) ?: return Result.success()
        val relay = container.settingsStore.settingsFlow.first().relay
        if (!relay.enabled || relay.baseUrl.isBlank() || relay.token.isBlank()) {
            AppLogger.relay("LinkPreviewWorker: relay not configured, skip note=$noteId")
            return Result.success()
        }

        // 抽 URL → 归一化（小写 host、去 trailing /、剥常见追踪参数）→ distinct → 截 5 条。
        // 之前出现"两条预览框，一条 404 一条正常"的根因：同一逻辑 URL 被解析成两个不同字符串
        // （比如带 / 不带 trailing slash、带 / 不带 utm），distinct 失效。
        val rawUrls = extractUrls(
            buildString {
                append(note.title).append('\n')
                append(note.content).append('\n')
                note.transcript?.let { append(it) }
            },
        )
        val urls = rawUrls.map { canonicalizeUrl(it) }
            .distinct()
            .take(MAX_URLS)
        AppLogger.relay(
            "LinkPreviewWorker: note=$noteId raw_urls=${rawUrls.size} dedup=${urls.size} list=${urls.joinToString(limit = 5)}",
        )
        if (urls.isEmpty()) {
            // 用户把 URL 删了又保存，旧的预览卡不清空会一直挂在笔记上。
            if (note.linkPreviews.isNotEmpty()) {
                container.noteRepository.setLinkPreviews(noteId, emptyList())
                AppLogger.relay("LinkPreviewWorker: note=$noteId cleared ${note.linkPreviews.size} stale previews")
            }
            return Result.success()
        }

        // 同一条笔记重复保存不必每次重抓：已经成功抓过且没有 error 的直接沿用。
        val existing = note.linkPreviews.associateBy { it.url }
        val now = System.currentTimeMillis()
        val previews = urls.map { url ->
            val cached = existing[url]
            if (cached != null && cached.error == null && cached.fetchedAt > 0 &&
                now - cached.fetchedAt < CACHE_TTL_MS
            ) {
                return@map cached
            }
            runCatching { container.linkPreviewClient.fetchPreview(relay, url) }
                .fold(
                    onSuccess = {
                        LinkPreview(
                            url = it.url.ifBlank { url },
                            title = it.title,
                            description = it.description,
                            imageUrl = it.imageUrl,
                            siteName = it.siteName,
                            summary = it.summary,
                            fetchedAt = now,
                            error = it.error,
                        )
                    },
                    onFailure = { e ->
                        LinkPreview(
                            url = url,
                            fetchedAt = now,
                            error = (e.message ?: "unknown").take(120),
                        )
                    },
                )
        }

        container.noteRepository.setLinkPreviews(noteId, previews)
        AppLogger.relay("LinkPreviewWorker: note=$noteId wrote ${previews.size} previews")
        return Result.success()
    }

    companion object {
        const val KEY_NOTE_ID = "note_id"
        private const val MAX_URLS = 5
        private const val CACHE_TTL_MS = 7L * 24 * 60 * 60 * 1000  // 7 天

        fun schedule(ctx: Context, noteId: String) {
            val req = OneTimeWorkRequestBuilder<LinkPreviewWorker>()
                .setInputData(workDataOf(KEY_NOTE_ID to noteId))
                .setConstraints(
                    Constraints.Builder()
                        .setRequiredNetworkType(NetworkType.CONNECTED)
                        .build(),
                )
                .setBackoffCriteria(BackoffPolicy.EXPONENTIAL, 30, TimeUnit.SECONDS)
                .build()
            WorkManager.getInstance(ctx).enqueueUniqueWork(
                "link_preview_$noteId",
                ExistingWorkPolicy.REPLACE,
                req,
            )
        }
    }
}

// URL 只允许 RFC 3986 合法字符（ASCII 内），这样遇到中文 / 全角标点 / 换行自动终止。
// 之前用 `[^\s]+` 的版本把 "https://x.com/y，这是一个..." 后面的中文都当 URL 吞进去了，
// 送到 relay 的 urllib 再报 'ascii' codec can't encode characters ... 的错。
private val URL_REGEX = Regex(
    """https?://[A-Za-z0-9\-._~:/?#\[\]@!$&'()*+,;=%]+""",
    RegexOption.IGNORE_CASE,
)

/**
 * 从文本里扫 URL。正则只吃 ASCII URL 合法字符，遇到中文立即终止；
 * 再剥一层行尾常见标点（兜底半角 . , 以及用户偶尔输入的全角符号），
 * 例如 "看这个 https://x.com/a." 的末尾句号不应该留在 URL 里。
 */
internal fun extractUrls(text: String): List<String> =
    URL_REGEX.findAll(text)
        .map { it.value.trimEnd('.', ',', ';', ':', '，', '。', '、', '；', '：', '）', ')', ']', '」', '』', '"', '\'') }
        .filter { it.length > 8 }  // http://x.y 以下长度忽略
        .toList()

// 常见的追踪参数：来自不同分享渠道的 utm_*、各家短链平台的 from/spm/share/scene/ref 等等。
// 笔记里同一篇 URL 在两次粘贴时一次带 utm 一次不带，会被算成两条不同 URL，
// 进而出现"一条 404、一条正常"或两个相同预览框的体验。
private val TRACKING_PARAM_PREFIXES = setOf(
    "utm_", "spm", "share", "from", "ref", "ref_", "scene", "src", "sourceid",
    "session_id", "weibo_id", "_t", "fr",
)

/**
 * 归一化 URL，便于跨形态去重：
 * - scheme/host 转小写；
 * - 去掉 trailing 斜杠（path == "/" 的根路径除外）；
 * - 去掉 query 中的常见追踪参数；
 * - fragment 保留（一些站点的 #id 是关键路由）。
 *
 * 解析失败时返回原串，至少不要让笔记保存链路 throw。
 */
internal fun canonicalizeUrl(raw: String): String {
    return runCatching {
        val parsed = java.net.URI(raw)
        val scheme = parsed.scheme?.lowercase() ?: return raw
        val host = parsed.host?.lowercase() ?: return raw
        val port = if (parsed.port == -1) "" else ":${parsed.port}"
        val rawPath = parsed.rawPath ?: ""
        val path = if (rawPath.length > 1 && rawPath.endsWith('/')) rawPath.trimEnd('/') else rawPath
        val cleanedQuery = parsed.rawQuery?.split('&')
            ?.filter { kv ->
                val key = kv.substringBefore('=').lowercase()
                key.isNotBlank() && TRACKING_PARAM_PREFIXES.none { prefix -> key == prefix || key.startsWith(prefix) }
            }
            ?.takeIf { it.isNotEmpty() }
            ?.joinToString("&")
        val frag = parsed.rawFragment?.takeIf { it.isNotBlank() }
        buildString {
            append(scheme).append("://").append(host).append(port).append(path)
            if (!cleanedQuery.isNullOrBlank()) append('?').append(cleanedQuery)
            if (frag != null) append('#').append(frag)
        }
    }.getOrDefault(raw)
}

File: relay_service/app/link_preview_worker.py
"""链接预览后台抓取（Web 对齐 Android v1.1.0 J Phase 1）。

笔记保存 / WebDAV pull 后 fire-and-forget 调用 schedule_for_note(note_id)：
1. 提取 content + title + transcript 中的 URL；
2. 用 canonicalize_url 归一化 + distinct + take 5；
3. 已成功抓过且 < 7 天 TTL 的复用旧 cache，不重抓；
4. 调 routes_links.fetch_preview_internal 拉新预览；
5. 写回 SQLite notes.link_previews_json；**不**触发 WebDAV 推（链接预览是 server-local cache，
   跨同步会让 Android 端反复重抓）。
"""

from __future__ import annotations

import asyncio
import json
import logging
import re
import time
import urllib.parse
from typing import Iterable

logger = logging.getLogger("link_preview")

MAX_URLS = 5
CACHE_TTL_MS = 7 * 24 * 60 * 60 * 1000  # 7 天

# RFC 3986 ASCII URL 字符集；遇到中文 / 全角符号自动终止。
URL_REGEX = re.compile(r"https?://[A-Za-z0-9\-._~:/?#\[\]@!$&'()*+,;=%]+", re.IGNORECASE)
TRAILING_PUNCT = ".,;:，。、；：）)]」』\"'"

# 与 Android LinkPreviewWorker.canonicalizeUrl 同集追踪参数前缀
TRACKING_PARAM_PREFIXES = (
    "utm_", "spm", "share", "from", "ref", "ref_", "scene", "src", "sourceid",
    "session_id", "weibo_id", "_t", "fr",
)


def extract_urls(text: str) -> list[str]:
    """从文本里扫 URL，剥行尾常见标点后返回"""
    out: list[str] = []
    if not text:
        return out
    for m in URL_REGEX.finditer(text):
        url = m.group(0)
        # 剥末尾标点（半角 + 全角）
        while url and url[-1] in TRAILING_PUNCT:
            url = url[:-1]
        if len(url) > 8:
            out.append(url)
    return out


def canonicalize_url(raw: str) -> str:
    """归一化 URL 用于 dedup：小写 host、去 trailing /、剥追踪参数。失败返回原串。

    与 Android LinkPreviewWorker.canonicalizeUrl 等价。"""
    try:
        parts = urllib.parse.urlsplit(raw)
        scheme = (parts.scheme or "").lower()
        host = (parts.hostname or "").lower()
        if not scheme or not host:
            return raw
        port = f":{parts.port}" if parts.port else ""
        # path: 去 trailing / 但保留根路径
        path = parts.path or ""
        if len(path) > 1 and path.endswith("/"):
            path = path.rstrip("/")
        # query: 剥追踪参数（按前缀匹配 + 完整匹配）
        query_pairs = []
        if parts.query:
            for kv in parts.query.split("&"):
                key = kv.split("=", 1)[0].lower()
                if not key:
                    continue
                if any(key == p or (p.endswith("_") and key.startswith(p)) or key.startswith(p) for p in TRACKING_PARAM_PREFIXES):
                    continue
                query_pairs.append(kv)
        cleaned_query = "&".join(query_pairs)
        # fragment 保留
        frag = parts.fragment or ""
        return urllib.parse.urlunsplit((scheme, host + port, path, cleaned_query, frag))
    except Exception:
        return raw


async def schedule_for_note(note_id: str) -> None:
    """fire-and-forget：起一个后台任务抓笔记里的链接预览。"""
    asyncio.create_task(_run(note_id))


async def _run(note_id: str) -> None:
    try:
        await _fetch_and_persist(note_id)
    except Exception as e:
        logger.warning("link preview worker failed note=%s: %s", note_id, e)


async def _fetch_and_persist(note_id: str) -> None:
    from .database import get_db
    from .routes_links import fetch_preview_internal

    db = await get_db()
    rows = await db.execute_fetchall(
        "SELECT title, content, transcript, link_previews_json FROM notes WHERE id = ?",
        [note_id],
    )
    if not rows:
        return
    row = rows[0]
    text = "\n".join(filter(None, [row["title"], row["content"], row["transcript"]]))

    raw_urls = extract_urls(text)
    canonical = []
    seen: set[str] = set()
    for u in (canonicalize_url(x) for x in raw_urls):
        if u not in seen:
            seen.add(u)
            canonical.append(u)
        if len(canonical) >= MAX_URLS:
            break

    if not canonical:
        # 用户把 URL 删了又保存，旧的预览要清空
        existing_raw = row["link_previews_json"]
        if existing_raw and existing_raw != "[]":
            await db.execute(
                "UPDATE notes SET link_previews_json = '[]' WHERE id = ?",
                [note_id],
            )
            await db.commit()
        return

    # 复用 cache：已存在且无 error 且 < 7 天的直接保留，不重抓。
    existing_by_url: dict[str, dict] = {}
    try:
        for it in json.loads(row["link_previews_json"] or "[]"):
            if isinstance(it, dict) and it.get("url"):
                existing_by_url[it["url"]] = it
    except Exception:
        pass

    now = int(time.time() * 1000)
    new_previews: list[dict] = []
    for url in canonical:
        cached = existing_by_url.get(url)
        if cached and not cached.get("error") and cached.get("fetched_at", 0) > 0 and now - cached["fetched_at"] < CACHE_TTL_MS:
            new_previews.append(cached)
            continue
        try:
            resp = await fetch_preview_internal(url, want_summary=True)
            new_previews.append({
                "url": resp.url or url,
                "title": resp.title,
                "description": resp.description,
                "image_url": resp.image_url,
                "site_name": resp.site_name,
                "summary": resp.summary,
                "fetched_at": now,
                "error": resp.error,
            })
        except Exception as e:
            new_previews.append({
                "url": url, "title": "", "description": "", "image_url": "",
                "site_name": "", "summary": "", "fetched_at": now,
                "error": str(e)[:160],
            })

    encoded = json.dumps(new_previews, ensure_ascii=False)
    # 故意不动 updated_at / status / last_synced_at——预览是后台副作用，
    # 跨 WebDAV 同步推它没意义（Android 端有自己的 LinkPreviewWorker 重抓）。
    await db.execute(
        "UPDATE notes SET link_previews_json = ? WHERE id = ?",
        [encoded, note_id],
    )
    await db.commit()
    logger.info("link preview note=%s wrote %d previews", note_id, len(new_previews))


def _filter_pairs(query: str) -> Iterable[str]:
    """utility 留作未来扩展，目前 canonicalize_url 内联了"""
    yield from (kv for kv in query.split("&") if kv)
